src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Debugging/DataTipRangeHandlerEndpoint.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.LanguageServer.EndpointContracts;
+using Microsoft.AspNetCore.Razor.LanguageServer.Hosting;
+using Microsoft.AspNetCore.Razor.Threading;
+using Microsoft.CodeAnalysis.Razor.DocumentMapping;
+using Microsoft.CodeAnalysis.Razor.Logging;
+using Microsoft.CodeAnalysis.Razor.Protocol;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Debugging;
+
+[RazorLanguageServerEndpoint(VSInternalMethods.TextDocumentDataTipRangeName)]
+internal class DataTipRangeHandlerEndpoint(
+    IDocumentMappingService documentMappingService,
+    LanguageServerFeatureOptions languageServerFeatureOptions,
+    IClientConnection clientConnection,
+    ILoggerFactory loggerFactory)
+    : AbstractRazorDelegatingEndpoint<TextDocumentPositionParams, VSInternalDataTip?>(
+        languageServerFeatureOptions,
+        documentMappingService,
+        clientConnection,
+        loggerFactory.GetOrCreateLogger<DataTipRangeHandlerEndpoint>()), ICapabilitiesProvider
+{
+    protected override bool OnlySingleServer => false;
+
+    protected override string CustomMessageTarget => CustomMessageNames.RazorDataTipRangeName;
+
+    public void ApplyCapabilities(VSInternalServerCapabilities serverCapabilities, VSInternalClientCapabilities clientCapabilities)
+    {
+        serverCapabilities.EnableDataTipRangeProvider();
+    }
+
+    protected override Task<IDelegatedParams?> CreateDelegatedParamsAsync(TextDocumentPositionParams request, RazorRequestContext requestContext, DocumentPositionInfo positionInfo, CancellationToken cancellationToken)
+    {
+        // only C# supports breakpoints
```

Comment:
>C# [](http://example.com/codeflow?start=16&length=2)

Technically JavaScript does as well, and is Roslyn-based. Not sure if it supports this specific capability though and HTML LSP wouldn't pass through to JavaScript without any extra work, so a moot point for now.

Reply:
I copied this comment from one of the other debugging endpoints. I don't mind changing it, but it feels like it's probably accurate for the foreseeable future as js debugging in razor in VS is incomplete.

Reply:
Fair enough.